src/nodes/display/ColorSpaceNode.js

```diff
@@ -78,8 +78,8 @@ export default ColorSpaceNode;
 
 ColorSpaceNode.type = /*@__PURE__*/ registerNode( 'ColorSpace', ColorSpaceNode );
 
-export const linearSRGBToColorSpace = ( node, colorSpace = null ) => nodeObject( new ColorSpaceNode( nodeObject( node ), colorSpace, LinearSRGBColorSpace ) );
-export const colorSpaceToLinearSRGB = ( node, colorSpace = null ) => nodeObject( new ColorSpaceNode( nodeObject( node ), LinearSRGBColorSpace, colorSpace ) );
+export const toOutputColorSpace = ( node, colorSpace = null ) => nodeObject( new ColorSpaceNode( nodeObject( node ), colorSpace, LinearSRGBColorSpace ) );
+export const toWorkingColorSpace = ( node, colorSpace = null ) => nodeObject( new ColorSpaceNode( nodeObject( node ), LinearSRGBColorSpace, colorSpace ) );
```

Comment:
These methods should not be conflated with `outputColorSpace`.

Reply:
@donmccurdy We need to be able to cover to log space in a shader, and then convert back.

```js
export const workingToColorSpace = ( node, colorSpace = null ) => nodeObject( new ColorSpaceNode( nodeObject( node ), colorSpace, workingColorSpace ) );

export const colorSpaceToWorking = ( node, colorSpace = null ) => nodeObject( new ColorSpaceNode( nodeObject( node ), workingColorSpace, colorSpace ) );
```

Reply:
`toOutputColorSpace` could be renamed to `toColorSpace`, but that begs the question: "from where?"

Reply:
`toOutputColorSpace()` will use `renderer.outputColorSpace` as default value, so I think there should be some representation. We have two base nomenclatures, which would be `Output Color Space` and `Working Color Space` in general terms.

Reply:
Another thing we could take advantage of from TSL is to inform through a context which `color space` that node is inserted in the code flow, this would avoid redundant conversions. Similar to what we used in `node.toTexture()` recently.

Reply:
That is fine. But you removed methods we need, which you can't do, unless you replace it with something equivalent.

Reply:
You can restore what was there before, or replace them with the two methods I [posted](https://github.com/mrdoob/three.js/pull/29259#discussion_r1736832008) above. Doing that later requires knowledge of the working color space.

Reply:
I think if you use toOutputColorSpace() it is already understood that the source color space is working color space. The same would happen in reverse, if you use toWorkingColorSpace() the source must be the output color space. I think for more specific cases this should be handled with another code and functions.

Reply:
Shader code prior to this PR:

```js
let c = colorNode.linearSRGBToColorSpace( SRGBColorSpace ).mul( _matrix ).colorSpaceToLinearSRGB( SRGBColorSpace );
```

You have removed those methods. With this PR, the following is required:

```js
let c = colorNode.toOutputColorSpace( SRGBColorSpace ).mul( _matrix ).toWorkingColorSpace( SRGBColorSpace );
```

1. `toOutputColorSpace` is misnamed. It has nothing to do with `outputColorSpace`. Rename to `toColorSpace` or `workingToColorSpace`.

2. `toOutputColorSpace` hard-wires the source color space to linear-sRGB. The source should be the working color space.

3. `toWorkingColorSpace` hard-wires the destination color space to linear-sRGB. The destination should be the working color space.

4. Please restore `linearSRGBToColorSpace` and `colorSpaceToLinearSRGB` -- at least for now.

Reply:
`OutputColorSpace` and `WorkerColorSpace` are for applications based on renderer settings, like `renderer.outputColorSpace` and do not need mandatory parameters:

```js
const c = colorNode.toOutputColorSpace().mul( _matrix ).toWorkingColorSpace();
```

Did you imagine that when the user needed to use another `renderer.outputColorSpace` like `DisplayP3ColorSpace` instead `SRGBColorSpace` he would have to readjust the code with conditionals if he were using `linearSRGBToColorSpace`?

The current abstraction should solves this issue.

I don't see a problem in exporting the most direct functions related to color space conversions compatible with tree-shaking, in fact this is good, `linearSRGBTosRGB()` and other functions are fine, but this would not cancel out the benefit of the current one.

Reply:
I do not object to any additional convenience methods you want to add.

The color space I need to covert to (and from) is a _specific_ color space -- _not related_ to the current render's output color space.

That is why, when we spoke offline, I suggested `linearSRGBToColorSpace( targetColorSpace )` -- or preferably, `workingToColorSpace( targetColorSpace )`.

Reply:
I'm happy with either of these parallel naming choices:

- workingToColorSpace / colorSpaceToWorking
- toWorkingColorSpace( sourceColorSpace ) / fromWorkingColorSpace( targetColorSpace )
- etc

Additional helper functions (e.g. related to the output color space) are certainly OK with me too.

Reply:
@sunag Can you please add the convenience functions we need
```js
export const toColorSpace ( targetColorSpace, sourceColorSpace ) // a general function

export const workingToColorSpace ( targetColorSpace )

export const colorSpaceToWorking ( sourceColorSpace )

export const linearSRGBToColorSpace ( targetColorSpace ) // if you want some specific ones

export const colorSpaceToLinearSRGB ( sourceColorSpace ) // if you want some specific ones
```

also, you seem to have hardwired the working color space to`LinearSRGBColorSpace`. Why did you do that?

```js
export const toOutputColorSpace = ( node, colorSpace = null ) => nodeObject( new ColorSpaceNode( nodeObject( node ), colorSpace, LinearSRGBColorSpace ) );
export const toWorkingColorSpace = ( node, colorSpace = null ) => nodeObject( new ColorSpaceNode( nodeObject( node ), LinearSRGBColorSpace, colorSpace ) );
```

Reply:
I'll look into adding `workingToColorSpace` and `colorSpaceToWorking`.

For `toColorSpace ( targetColorSpace, sourceColorSpace )` we should use direct functions like [these](https://github.com/mrdoob/three.js/blob/dev/src/nodes/display/ColorSpaceFunctions.js) `sRGBToLinearSRGB`, `linearSRGBTosRGB`, ...

Just for us to think about, we shouldn't get used to string as a parameter, because that requires us to create a library to support it and that breaks tree-shaking.

If we have 4 different types of color-space between working/output, we will have to have 16 conversion functions and this growth is exponential for each additional color-space that we add here.

The future of `renderer.outputColorSpace` type could be a Node instead of string, this is one of the valid reasons for the current abstraction output/working names.

That's not a concern right now, I just wanted to clarify those points.

Reply:
@sunag I share your concern about the growing number of color space pairs, and their effect on tree-shaking. But I don't think I agree about using direct 1-to-1 functions or nodes for the conversions. For one thing, we have to import those functions, and call them conditionally, and if we have 4x4=16 functions it's going to be hard to ensure they're tree-shakeable for internal renderer usage.

A solution I'd prefer would be to continue using the predefined color space strings from WebGPU and CSS Color Module 4, but to allow users to define their own, and to import color spaces from three/addons. For example:

```javascript
import { ColorManagement } from 'three';
import { Rec2100HLGColorSpace, Rec2100HLGDefinition } from 'three/addons/math/colorspaces/Rec2100HLG.js';

ColorManagement.addColorSpace( Rec2100HLGColorSpace, Rec2100HLGDefinition );
```

Each definition would be similar to what you see in [`ColorManagement.js`](https://github.com/mrdoob/three.js/blob/c85a4374c15aa6aeff6768fd3678eead19ebd68f/src/math/ColorManagement.js#L28-L61), but represented as data properties, not JavaScript, so it's easy to do conversions in JS, GLSL, or WGSL. There's a similar approach in nanocolor (C/C++) which I believe is well done: https://github.com/meshula/Nanocolor/blob/main/nanocolor.c#L150

Once registered, a color space could be used for any compatible purpose. Any color space conversions to/from a color space not registered with `THREE.ColorManagement` may be considered an error. We could probably drop Display P3 and Linear Display P3 from the core library.

Reply:
The conversion solution for direct functions is just the suggested function `toColorSpace()`, where it would be cleaner to do `sRGBToLinearSRGB( a )` than `a.toColorSpace( THREE.SRGBColorSpace, THREE.LinearSRGBColorSpace )` apart from the benefit of tree-shaking.

I don't think there was an appropriate solution in this sense for working/output so I created the PRs `workingToColorSpace` and `colorSpaceToWorking` following the suggestions.

I created a system based on this principle, it should already be functional. I think we can encourage users to add functions like a library to their projects instead of adding them all to the core, it would be a solution.

- StandardNodeLibrary
https://github.com/mrdoob/three.js/blob/f509e7bf0ae31a83cd30f6cd8cf269f3708dc6f2/src/renderers/webgpu/nodes/StandardNodeLibrary.js#L98-L99

This could be added externally with:
```js
const nodeLibrary = renderer.nodes.library;

nodeLibrary.addColorSpace( linearSRGBTosRGB, getColorSpaceMethod( LinearSRGBColorSpace, SRGBColorSpace ) ); 
nodeLibrary.addColorSpace( sRGBToLinearSRGB, getColorSpaceMethod( SRGBColorSpace, LinearSRGBColorSpace ) );
```

Reply:
I don't think we can ask users to foresee all possible source/target color space pairs, though. And with only a function/node defining a color space, it won't be possible to:

1. Apply CPU-side calculations, like converting `new Color( 0x808080 )` from sRGB to a user-defined working color space
2. Determine whether a texture in a user-defined color space requires unpacking with gl.unpackColorSpace or not. For this we need to be able to query, for a particular texture's color space, what are the associated transfer functions?

I think the `workingToColorSpace` and `colorSpaceToWorking` functions will fit into either solution well. But I would like to refactor THREE.ColorManagement a little, so that the node system can infer how to convert between two previously unknown color spaces based on their definitions (primaries, white point, transfer functions), without needing to have a conversion function/node registered in advance.

Reply:
> I don't think we can ask users to foresee all possible source/target color space pairs, though. And with only a function/node defining a color space, it won't be possible to:

The idea of ​​the user adding other ColorSpaces would be related to the colorspaces currently in effect in the project or in your library, not in a global sense.
 
 It would be great to have your collaboration in this regard :)

Reply:
Continued in https://github.com/mrdoob/three.js/pull/29450.